### ISSUE TITLE
cache_video failed, error: result type Float can't be cast to the desired output type Byte

### ISSUE THREAD

**Issue by Tan-Junwen** [Westlake-AGI-Lab/FlowDirector#1]
[2025-07-01 07:17:21,561] INFO: Generation model config: {'__name__': 'Config: Wan T2V 1.3B', 't5_model': 'umt5_xxl', 't5_dtyp
e': torch.bfloat16, 'text_len': 512, 'param_dtype': torch.bfloat16, 'num_train_timesteps': 1000, 'sample_fps': 16, 'sample_neg
_prompt': '色调艳丽，过曝，静态，细节模糊不清，字幕，风格，作品，画作，画面，静止，整体发灰，最差质量，低质量，JPEG压缩残留， 
丑陋的，残缺的，多余的手指，画得不好的手部，画得不好的脸部，畸形的，毁容的，形态畸形的肢体，手指融合，静止不动的画面，杂乱的背
景，三条腿，背景人很多，倒着走', 't5_checkpoint': 'models_t5_umt5-xxl-enc-bf16.pth', 't5_tokenizer': 'google/umt5-xxl', 'vae_c
heckpoint': 'Wan2.1_VAE.pth', 'vae_stride': (4, 8, 8), 'patch_size': (1, 2, 2), 'dim': 1536, 'ffn_dim': 8960, 'freq_dim': 256,
 'num_heads': 12, 'num_layers': 30, 'window_size': (-1, -1), 'qk_norm': True, 'cross_attn_norm': True, 'eps': 1e-06}          
[2025-07-01 07:17:21,561] INFO: Input prompt: A large dinosaur is walking slowly across a rocky terrain in a zoo enclosure, su
rrounded by stone walls and scattered greenery. The camera remains fixed, capturing the dinosaur's deliberate movements.      
[2025-07-01 07:17:21,561] INFO: Creating WanT2V pipeline.                                                                     
[2025-07-01 07:18:22,293] INFO: loading /fsave/tjw/FlowDirector/Wan2.1-T2V-1.3B/models_t5_umt5-xxl-enc-bf16.pth               
[2025-07-01 07:18:36,441] INFO: loading /fsave/tjw/FlowDirector/Wan2.1-T2V-1.3B/Wan2.1_VAE.pth                                
[2025-07-01 07:18:36,732] INFO: Creating WanModel from /fsave/tjw/FlowDirector/Wan2.1-T2V-1.3B                                
[2025-07-01 07:18:42,869] INFO: Generating video ...                                                                          
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████
████████████████████████████████████████████████| 50/50 [09:23<00:00, 11.27s/it]                                              
[2025-07-01 07:28:21,586] INFO: Saving generated video to videos/0701_0728_large_brown_bear_large_dinosaur_n4_w2_omega2.75_s42
.mp4                                                                                                                          
cache_video failed, error: result type Float can't be cast to the desired output type Byte                                    
[2025-07-01 07:28:22,575] INFO: Finished.

**Comment by ZeChen-AI** [comment]
@MathamPollard  I replace the `cache_video` function in `wan/utils/utils.py `with the implementation below, it works.
```
def cache_video(tensor,
                save_file=None,
                fps=30,
                suffix='.mp4',
                nrow=8,
                normalize=True,
                value_range=(-1, 1),
                retry=5):
    # cache file
    cache_file = osp.join('/tmp', rand_name(
        suffix=suffix)) if save_file is None else save_file

    # save to cache
    error = None
    for _ in range(retry):
        try:
            # Ensure tensor is float32 and on CPU for processing
            tensor = tensor.float().cpu()
            
            # Comprehensive numerical safety checks
            def make_numerically_safe(t):
                # Replace inf and nan
                t = torch.nan_to_num(t, nan=0.0, posinf=1.0, neginf=-1.0)
                # Remove extreme values
                t = t.clamp(-10.0, 10.0)
                return t
            
            # Apply safety checks
            tensor = make_numerically_safe(tensor)
            
            # Apply initial clamp
            tensor = tensor.clamp(min(value_range), max(value_range))
            
            # Process each frame individually to avoid issues with make_grid
            frames = []
            
            # Check tensor dimensions and normalize to [B, C, T, H, W] format
            if tensor.ndim == 4:  # [C, T, H, W] - add batch dimension
                tensor = tensor.unsqueeze(0)  # [1, C, T, H, W]
            elif tensor.ndim == 5:  # [B, C, T, H, W] - already correct
                pass
            else:
                raise ValueError(f"Unsupported tensor shape: {tensor.shape}. Expected 4D [C,T,H,W] or 5D [B,C,T,H,W]")
            
            batch_size, channels, time_frames, height, width = tensor.shape
            
            for frame_idx in range(time_frames):
                frame = tensor[:, :, frame_idx, :, :]  # [B, C, H, W]
                
                # Apply safety checks to each frame
                frame = make_numerically_safe(frame)
                
                # Use make_grid with additional safety
                try:
                    grid = torchvision.utils.make_grid(
                        frame, nrow=nrow, normalize=normalize, value_range=value_range)
                    
                    # Apply safety checks after make_grid
                    grid = make_numerically_safe(grid)
                    
                    # Ensure values are in [0, 1] range
                    if normalize:
                        grid = (grid - grid.min()) / (grid.max() - grid.min() + 1e-8)
                    else:
                        grid = (grid - min(value_range)) / (max(value_range) - min(value_range))
                    
                    grid = grid.clamp(0.0, 1.0)
                    frames.append(grid)
                    
                except Exception as e:
                    print(f"Warning: make_grid failed for frame {frame_idx}, using fallback method: {e}")
                    # Fallback: simple normalization
                    grid = frame.mean(dim=0)  # Average over batch if needed
                    grid = (grid - grid.min()) / (grid.max() - grid.min() + 1e-8)
                    grid = grid.clamp(0.0, 1.0)
                    frames.append(grid)
            
            # Stack frames and permute dimensions
            tensor = torch.stack(frames, dim=0).permute(0, 2, 3, 1)  # [T, H, W, C]
            
            # Final safety check and conversion to uint8
            tensor = make_numerically_safe(tensor)
            tensor = tensor.clamp(0.0, 1.0)
            
            # Convert to uint8 with explicit safety
            tensor_uint8 = (tensor * 255.0).round().clamp(0, 255)
            
            # Check if conversion is safe
            if torch.isinf(tensor_uint8).any() or torch.isnan(tensor_uint8).any():
                print("Warning: Found inf/nan after uint8 conversion, applying final correction")
                tensor_uint8 = torch.nan_to_num(tensor_uint8, nan=0.0, posinf=255.0, neginf=0.0)
                tensor_uint8 = tensor_uint8.clamp(0, 255)
            
            tensor_uint8 = tensor_uint8.to(torch.uint8)

            # write video
            writer = imageio.get_writer(
                cache_file, fps=fps, codec='libx264', quality=8)
            for frame in tensor_uint8.numpy():
                writer.append_data(frame)
            writer.close()
            return cache_file
        except Exception as e:
            error = e
            print(f"cache_video attempt failed: {error}")
            continue
    else:
        print(f'cache_video failed after {retry} attempts, error: {error}', flush=True)
        return None
```

**Comment by Tan-Junwen** [comment]
It works! Thank you!
